freqtrade/rpc/api_server/api_ws.py

```diff
@@ -74,65 +77,28 @@ async def _process_consumer_request(
 
         # Format response
         response = WSWhitelistMessage(data=whitelist)
-        # Send it back
-        await channel_manager.send_direct(channel, response.dict(exclude_none=True))
+        await channel.send(response.dict(exclude_none=True))
 
     elif type == RPCRequestType.ANALYZED_DF:
-        limit = None
-
-        if data:
-            # Limit the amount of candles per dataframe to 'limit' or 1500
-            limit = max(data.get('limit', 1500), 1500)
+        # Limit the amount of candles per dataframe to 'limit' or 1500
+        limit = min(data.get('limit', 1500), 1500) if data else None
 
         # For every pair in the generator, send a separate message
         for message in rpc._ws_request_analyzed_df(limit):
+            # Format response
             response = WSAnalyzedDFMessage(data=message)
-            await channel_manager.send_direct(channel, response.dict(exclude_none=True))
+            await channel.send(response.dict(exclude_none=True))
 
 
 @router.websocket("/message/ws")
 async def message_endpoint(
-    ws: WebSocket,
+    websocket: WebSocket,
+    token: str = Depends(validate_ws_token),
     rpc: RPC = Depends(get_rpc),
-    channel_manager=Depends(get_channel_manager),
-    token: str = Depends(validate_ws_token)
+    message_stream: MessageStream = Depends(get_message_stream)
 ):
-    """
-    Message WebSocket endpoint, facilitates sending RPC messages
-    """
-    try:
-        channel = await channel_manager.on_connect(ws)
-        if await is_websocket_alive(ws):
-
-            logger.info(f"Consumer connected - {channel}")
-
-            # Keep connection open until explicitly closed, and process requests
-            try:
-                while not channel.is_closed():
-                    request = await channel.recv()
-
-                    # Process the request here
-                    await _process_consumer_request(request, channel, rpc, channel_manager)
-
-            except (WebSocketDisconnect, WebSocketException):
-                # Handle client disconnects
-                logger.info(f"Consumer disconnected - {channel}")
-            except RuntimeError:
-                # Handle cases like -
-                # RuntimeError('Cannot call "send" once a closed message has been sent')
-                pass
-            except Exception as e:
-                logger.info(f"Consumer connection failed - {channel}: {e}")
-                logger.debug(e, exc_info=e)
-
-    except RuntimeError:
-        # WebSocket was closed
-        # Do nothing
-        pass
-    except Exception as e:
-        logger.error(f"Failed to serve - {ws.client}")
-        # Log tracebacks to keep track of what errors are happening
-        logger.exception(e)
-    finally:
-        if channel:
-            await channel_manager.on_disconnect(ws)
+    async with create_channel(websocket) as channel:
```

Comment:
There's still a PRETTY nasty exception if i connect to a websocket with an invalid token.
it's curious though - as in my understanding - with an invalid token, we shouldn't even get this far into the method (so line 100 shouldn't be reached).

i can however 100% reliably reproduce this by simply using a wrong token on the consumer.

```
2022-11-24 07:09:46,596 - uvicorn.error - INFO - ('127.0.0.1', 47896) - "WebSocket /api/v1/message/ws?token=secret_ws_t0ken.1" 403
2022-11-24 07:09:46,596 - uvicorn.error - INFO - connection failed (403 Forbidden)                    2022-11-24 07:09:46,598 - uvicorn.error - INFO - connection closed
2022-11-24 07:09:46,598 - freqtrade.rpc.api_server.ws.channel - INFO - Disconnected from channel - WebSocketChannel(3ba2829a, ('127.0.0.1', 47896))                                                         2022-11-24 07:09:46,599 - uvicorn.error - ERROR - Exception in ASGI application
Traceback (most recent call last):
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/uvicorn/protocols/websockets/websockets_impl.py", line 230, in run_asgi
    result = await self.app(self.scope, self.asgi_receive, self.asgi_send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__                                                               return await self.app(scope, receive, send)                                                         File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/uvicorn/middleware/message_logger.py", line 86, in __call__                                                              raise exc from None
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/uvicorn/middleware/message_logger.py", line 82, in __call__
    await self.app(scope, inner_receive, inner_send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/fastapi/applications.py", line 270, in __call__                                                                          await super().__call__(scope, receive, send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__                                                                        await self.middleware_stack(scope, receive, send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/middleware/errors.py", line 149, in __call__                                                                   await self.app(scope, receive, send)                                                                File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/middleware/cors.py", line 76, in __call__
    await self.app(scope, receive, send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/routing.py", line 706, in __call__
    await route.handle(scope, receive, send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/routing.py", line 341, in handle
    await self.app(scope, receive, send)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/routing.py", line 82, in app
    await func(session)
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/fastapi/routing.py", line 287, in app
    await dependant.call(**values)
  File "/home/matt/freqtrade/freqtrade/rpc/api_server/api_ws.py", line 100, in message_endpoint
    async with create_channel(websocket) as channel:
  File "/home/matt/.pyenv/versions/3.10.4/lib/python3.10/contextlib.py", line 199, in __aenter__
    return await anext(self.gen)
  File "/home/matt/freqtrade/freqtrade/rpc/api_server/ws/channel.py", line 231, in create_channel
    await channel.accept()
  File "/home/matt/freqtrade/freqtrade/rpc/api_server/ws/channel.py", line 130, in accept
    return await self._websocket.accept()
  File "/home/matt/freqtrade/freqtrade/rpc/api_server/ws/proxy.py", line 73, in accept
    return await self._websocket.accept()
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/websockets.py", line 99, in accept
    await self.send(
  File "/home/matt/.pyenv/versions/3.10.4/envs/trade_3104/lib/python3.10/site-packages/starlette/websockets.py", line 87, in send
    raise RuntimeError('Cannot call "send" once a close message has been sent.')
RuntimeError: Cannot call "send" once a close message has been sent.
2022-11-24 07:09:51,066 - DummyDataStrategyTracker - INFO - RAM: 332.82 MB
```

Reply:
So the denying because of invalid token is a little different than a regular endpoint. The actual endpoint function is still run, but when the invalid token is detected it has simply closed the connection already. That's where the RuntimeError comes in from sending an "accept" message when the connection has already sent a "closed" message. This is fixed in the most recent commit [101dec4](https://github.com/freqtrade/freqtrade/pull/7771/commits/101dec461e40c2b8ed15a7075bb4b7dc9099c7b2).

Reply:
Is there a reason that the log message when a websocket is denied `debug` level? I'd imagine it would be best as `info` as users should know when they're getting hit with incorrect tokens.

Reply:
Shouldn't we then check on `if token:`  and skip the whole body if it's not provided?

Now how you log a denied is ... questionable.
you already get an implict error from uvicorn - so i think we put it to debug as otherwise, we'd have 3 log-lines for one bad connect (connect, denied, (both from uvicorn) and "our denied").

Reply:
I think just marking the channel closed if unable to connect would be the way to go as it keeps things simple and readable, while accomplishing the same thing.

Reply:
> Now how you log a denied is ... questionable. you already get an implict error from uvicorn - so i think we put it to debug as otherwise, we'd have 3 log-lines for one bad connect (connect, denied, (both from uvicorn) and "our denied").

I think in this case we might as well just remove that call altogether.